lib/ts/recipe/multifactorauth/recipe.ts

```diff
@@ -0,0 +1,508 @@
+            factorsSetUpForUser,
+            requiredSecondaryFactorsForTenant: tenantInfo.requiredSecondaryFactors ?? [],
+            requiredSecondaryFactorsForUser,
+            completedFactors: completedFactorsClaimValue?.c ?? {},
+            userContext: userContext,
+        });
+
+        await this.recipeInterfaceImpl.checkAllowedToSetupFactorElseThrowInvalidClaimError({
+            session: session,
+            factorId: factorId,
+            completedFactors: completedFactorsClaimValue?.c ?? {},
+            requiredSecondaryFactorsForTenant: tenantInfo.requiredSecondaryFactors ?? [],
+            requiredSecondaryFactorsForUser,
+            factorsSetUpForUser,
+            mfaRequirementsForAuth,
+            userContext: userContext,
+        });
+    };
+
+    checkFactorUserAccountInfoForVerification = (
+        sessionUser: User,
+        accountInfo: { email?: string; phoneNumber?: string }
+    ): { status: "OK" } | { status: "VALIDATION_ERROR"; reason: string } => {
+        /*
+            We discussed another method but did not go ahead with it, details below:
+
+            We can allow the second factor to be linked to first factor even if the emails are different 
+            and not verified as long as there is no other user that exists (recipe or primary) that has the 
+            same email as that of the second factor. For example, if first factor is google login with e1 
+            and second factor is email password with e2, we allow linking them as long as there is no other 
+            user with email e2.
+
+            We rejected this idea cause if auto account linking is switched off, then someone else can sign up 
+            with google using e2. This is OK as it would not link (since account linking is switched off). 
+            But, then if account linking is switched on, then the google sign in (and not sign up) with e2 
+            would actually cause it to be linked with the e1 account.
+        */
+
+        // we allow setup of unverified account info only if the session user has the same account info
+        // and it is verified
+
+        if (accountInfo.email !== undefined) {
+            let foundVerifiedEmail = false;
+            for (const lM of sessionUser?.loginMethods) {
+                if (lM.email === accountInfo.email && lM.verified) {
+                    foundVerifiedEmail = true;
+                    break;
+                }
+            }
+            if (!foundVerifiedEmail) {
+                return {
+                    status: "VALIDATION_ERROR",
+                    reason: "Cannot complete MFA because of security reasons. Please contact support. (ERR_CODE_010)",
+                };
+            }
+        }
+
+        if (accountInfo.phoneNumber !== undefined) {
+            let foundVerifiedPhoneNumber = false;
+            for (const lM of sessionUser?.loginMethods) {
+                if (lM.phoneNumber === accountInfo.phoneNumber) {
```

Comment:
use samePhoneNumberAs function, and not ===

Reply:
Done